honeybee/face.py

```diff
@@ -1,158 +1,875 @@
+    """A single planar face.
+
+    Properties:
+        name
+        name_original
+        geometry
+        punched_geometry
+        vertices
+        punched_vertices
+        upper_left_vertices
+        normal
+        center
+        area
+        perimeter
+        type
+        boundary_condition
+        apertures
+        doors
+        parent
+        has_parent
+    """
+    TYPES = face_types
+    __slots__ = ('_name', '_name_original', '_geometry', '_parent', '_punched_geometry',
+                 '_apertures', '_doors', '_type', '_boundary_condition', '_properties')
 
-    def __init__(self, name, geometry, face_type=None, boundary_condition=None):
+    def __init__(self, name, geometry, type=None, boundary_condition=None):
         """A single planar face.
+
         Args:
-            name: Face name.
+            name: Face name. Must be < 100 characters.
             geometry: A ladybug-geometry Face3D.
-            face_type: Face type (e.g. wall, floor).
+            type: Face type. Default varies depending on the direction that
+                the Face geometry is points.
+                RoofCeiling = pointing upward within 30 degrees
+                Wall = oriented vertically within +/- 60 degrees
+                Floor = pointing downward within 30 degrees
             boundary_condition: Face boundary condition (Outdoors, Ground, etc.)
+                Default is Outdoors unless the center of the input geometry lies
+                below the XY plane, in which case it will be set to Ground.
         """
-        self.name = name
+        # process the name
+        self._name = valid_string(name, 'honeybee face name')
+        self._name_original = name
+
+        # process the geometry
         assert isinstance(geometry, Face3D), \
-            'Expected ladybug_geometry Face3D not {}'.format(type(geometry))
+            'Expected ladybug_geometry Face3D. Got {}'.format(type(geometry))
         self._geometry = geometry
-        # _parent will be set when the Face is added to a Zone
-        # in case of aperture it will be added when aperture is added to a Face.
-        self._parent = None
-        # get face type based on normal if face_type is not provided
-        face_type = face_type or get_type_from_normal(geometry.normal)
-        self._properties = Properties(face_type, boundary_condition)
-        self._writer = writer
+        self._parent = None  # _parent will be set when the Face is added to a Room
+        # initialize with no apertures/doors (they can be assigned later)
+        self._punched_geometry = None
         self._apertures = []
+        self._doors = []
+
+        # set face type based on normal if not provided
+        self.type = type or get_type_from_normal(geometry.normal)
+
+        # set boundary condition by the relation to ground plane if not provided
+        self.boundary_condition = boundary_condition or \
+            get_bc_from_position(geometry.center)
```

Comment:
Ok, I can change it to only assign the ground bc if all positions are at or below the ground depth.  I think the docstring changed from when you wrote this comment since it doesn't say that you can change the ground depth any more.  And this makes sense  given that, on the face-level, you can just input a Ground bc on `__init__`.  However, when you make a Room from a polyface, there's an input argument to change the ground plane.

Reply:
>  I think the docstring changed from when you wrote this comment since it doesn't say that you can change the ground depth any more.

See line 79. I meant the comment in the code not the docstring.

This has been one of those issues that comes up once in a while when someone is modeling a geometry on a slope, etc. I think this is an issue that we can revisit later.

Reply:
I made the comment more explicit that it is a zero ground plane by default. And I agree that checking all of the vertices is better practice.